### PR TITLE
MGDOBR-270: Event Sources business logic

### PR DESCRIPTION
**Pull request by kostola**
[MGDOBR-270](https://issues.redhat.com/browse/MGDOBR-270)

**Blocked by:**
- #662 
- #663 

## Description

This PR belongs to the [Event Sources epic](https://issues.redhat.com/browse/MGDOBR-270) and is part of a sequence of PRs derived from the Event Sources PoC (#629)

This PR is the *final one* of the sequence. It contains the business logic that will make the first version of the *Event Sources* feature actually usable.

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [x] Check if code and Github action workflows have been modified and if the modification is safe
- [x] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorsService.java [hunk 1]
```diff
@@ -30,10 +26,9 @@ public class ExecutorsService {
     @Incoming(EVENTS_IN_CHANNEL)
     public CompletionStage<Void> processBridgeEvent(final Message<String> message) {
         try {
-            CloudEvent cloudEvent = CloudEventUtils.decode(message.getPayload());
             Executor executor = executorsProvider.getExecutor();
             try {
-                executor.onEvent(cloudEvent);
+                executor.onEvent(message.getPayload());
```
Comment (danielezonca) [inline executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorsService.java:31]: We should move in the opposite direction: I would add the CloudEvents parsing here and preserve `executor.onEvent(CloudEvents)` method.
The whole system is designed to manage CloudEvents and in the future RHOC is going to do the conversion for us so at that point we will probably going to have only CloudEvents everywhere.

My suggestion is to check if this a valid CloudEvents, if not, "convert" it as CloudEvents (adding missing metadata) here

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java [hunk 2]
```diff
@@ -193,4 +193,14 @@ protected ConnectorEntity doDeleteDependencies(ConnectorEntity connectorEntity)
         return connectorEntity;
     }
 
+    private static boolean isSourceConnector(ConnectorEntity connectorEntity) {
+        return connectorEntity.getConnectorType().contains("_source_");
```
Comment (danielezonca) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java:197]: Can we use a proper `type` check instead of a contains?

File: manager/src/test/java/com/redhat/service/smartevents/manager/connectors/ConnectorsServiceTest.java [hunk 3]
```diff
@@ -33,77 +44,85 @@ class ConnectorsServiceTest {
...
+        }
+
         processor.setId(TEST_PROCESSOR_ID);
         processor.setName(TEST_PROCESSOR_NAME);
+        processor.setDefinition(processorDefinition);
         return processor;
     }
 
-    private Action testSlackAction() {
+    private static Action testSlackAction() {
```
Comment (danielezonca) [inline manager/src/test/java/com/redhat/service/smartevents/manager/connectors/ConnectorsServiceTest.java:125]: Minor note (and it is already on main so not introduced by this PR): I find confusing having a method named `test` if it is not a test :)

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        String endpoint = getOrThrowException(action, ENDPOINT_PARAM);
+        String cloudEventType = getOrThrowException(action, CLOUD_EVENT_TYPE_PARAM);
+        AbstractOidcClient abstractOidcClient = getOidcClient();
+        return new SourceActionInvoker(endpoint, cloudEventType, webClient, abstractOidcClient);
+    }
+
+    private static String getOrThrowException(Action action, String parameterName) {
+        if (action.getParameters() == null || !action.getParameters().containsKey(parameterName)) {
+            throw new GatewayProviderException(String.format("Missing %s parameter from action", parameterName));
+        }
```
Comment (danielezonca) [inline processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java:26]: Is this validation also performed during Action validation?

**Review by danielezonca — COMMENTED** [review]
Not a full review, just some comments :)

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        this.endpoint = endpoint;
+        this.cloudEventType = cloudEventType;
+        this.client = client;
+        this.oidcClient = oidcClient;
+        this.idGenerator = idGenerator;
+    }
+
+    @Override
+    public void onEvent(String event) {
+        LOG.info("SourceActionInvoker::onEvent | {} - {} | {}", endpoint, cloudEventType, event);
```
Comment (robpblake) [inline processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java:51]: @kostola Lots of logging in here that I think is a bit noisy and is also printing senstive data into the logs i.e the token. Can we tidy it up a bit please.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java [hunk 6]
```diff
@@ -302,16 +317,4 @@ public ProcessorResponse toResponse(Processor processor) {
 
         return processorResponse;
     }
-
-    JsonNode definitionToJsonNode(ProcessorDefinition definition) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:306]: @kostola Does the change in serialization/deserialization impact existing entities in the database? I.e when we deploy this to `dev`, will it fail to deserialize existing entities?

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            request = request.bearerTokenAuthentication(token);
+        }
+        request.sendJsonObject(getPayload(event)).subscribe().with(
+                response -> LOG.info("RESPONSE {} {}", response.statusCode(), response.statusMessage()),
+                error -> LOG.error("ERROR", error));
+    }
+
+    private JsonObject getPayload(String event) {
+        try {
+            ObjectNode cloudEvent = MAPPER.createObjectNode();
```
Comment (robpblake) [inline processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java:67]: @kostola Do you have an example of the Event put onto the Kafka Topic by the Slack Source Connector?

I'm interested to understand if the Event contains headers, because at the minute we're putting the Kafka Event headers into the `data` part of the `CloudEvent`. So from a user experience perspective, if I want to match on a header from a Source, I'm interested to understand if they have to write a filter that targets the `data` element of the `CloudEvent`.

e.g. if the Kafka Event produced looks like this:

```
{
 "header_1":"foo"
}
```
Do I as a user have to write the following filter to match on it:

```
{
  "filters": [
    {
      "type": "StringEquals", 
      "key": "data.header_1",
      "value": "foo"
    }
  ]
}
```

Does that match user expectations or would you as a user expect the headers to be hoisted as custom attributes on the `CloudEvent`? 🤔

File: executor/src/main/java/com/redhat/service/smartevents/executor/Executor.java [hunk 8]
```diff
@@ -50,31 +54,41 @@ public Executor(ProcessorDTO processor, FilterEvaluatorFactory filterEvaluatorFa
...
             // DO NOTHING;
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> toEventMap(String event) {
+        try {
+            return mapper.readValue(event, Map.class);
+        } catch (JsonProcessingException e) {
+            LOG.error("JsonProcessingException when generating event map for '{}'", event, e);
```
[the remark this answers is not in the document]
Comment (robpblake) [reply executor/src/main/java/com/redhat/service/smartevents/executor/Executor.java:87]: @jomarko Yes it's correct. If a `Throwable` is the last parameter of a log statement, SLF4j will print a stacktrace.

[the remark this answers is not in the document]
Comment (robpblake) [reply executor/src/main/java/com/redhat/service/smartevents/executor/Executor.java:87]: @jomarko Yes it's correct. If a `Throwable` is the last parameter of a log statement, SLF4j will print a stacktrace.

File: manager/src/test/java/com/redhat/service/smartevents/manager/connectors/ConnectorsServiceTest.java (hunk 3 above)
Comment (kostola) [reply manager/src/test/java/com/redhat/service/smartevents/manager/connectors/ConnectorsServiceTest.java]: I renamed those methods to avoid confusion :+1:

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java (hunk 4 above)
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java:26]: There's no action validation phase for `SourceAction` because it's an internal action used only by source processors. However, the existence of those parameters should be guaranteed by the unit tests of the source resolvers.

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java (hunk 5 above)
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: Cleaned this part :+1:

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Override
+    public void onEvent(String event) {
+        LOG.info("SourceActionInvoker::onEvent | {} - {} | {}", endpoint, cloudEventType, event);
+
+        HttpRequest<Buffer> request = client.postAbs(endpoint);
+        String token = oidcClient.getToken();
+        LOG.info("SourceActionInvoker::onEvent | Token: {}", token);
+        if (token != null && !"".equals(token)) {
```
[the remark this answers is not in the document]
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: Done :+1:

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java (hunk 6 above)
Comment (kostola) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:306]: Nope it should work because the column is always `jsonb` and the payload is compatible with the `ProcessorDefinition` bean

Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:306]: Thanks!

**Comment by manstis** [comment]
@kostola I will look at this tomorrow.

**Comment by manstis** [comment]
@kostola I notice there are some significant comments from @robpblake and @danielezonca.

Therefore rather than add more opinion to those already made; I'll defer my review until later (if needed).

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java (hunk 2 above)
Comment (kostola) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java]: Done :+1:

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java (hunk 4 above)
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvokerBuilder.java]: With the newest refactoring `SourceAction` is not needed anymore and standard `WebhookAction` can be used instead. Thus, I removed all this code.

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        LOG.debug("Received event with id '{}' and type '{}' in processor with name '{}' of bridge '{}", event.getId(), event.getType(), processor.getName(), processor.getBridgeId());
+
+        // Filter evaluation
+        if (!matchesFilters(eventMap)) {
+            LOG.debug("Filters of processor '{}' did not match for event with id '{}' and type '{}'", processor.getId(), event.getId(), event.getType());
+            return;
+        }
+        LOG.info("Filters of processor '{}' matched for event with id '{}' and type '{}'", processor.getId(), event.getId(), event.getType());
+        // Transformation
+        String eventToSend = applyTransformations(eventMap);
```
Comment (robpblake) [inline executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java:78]: @kostola For a `Source` processor, how do we guarantee that this transforms into a `CloudEvent`? From what I can see we simply send whatever the customer transforms into a JSON payload and send it to the Ingress via Webhook Action. Am I correct?

I think we have two options here:

- If the Processor is of type `Source`, then always marshall the transformed event back into a CloudEvent format
- Prevent the user from applying a transformation to a `Source` processor. As discussed before, that is out-of-scope for this work.

cheers,

Rob

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorService.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    @Inject
+    Executor executor;
+
+    @Inject
+    ObjectMapper mapper;
+
+    private Function<String, CloudEvent> toCloudEvent;
+
+    @PostConstruct
+    void init() {
```
Comment (robpblake) [inline executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorService.java:47]: @kostola Any reason not to just in-line this in the `processEvent` method? Avoids anyone having to remember to call the `init` method. 

Not a blocking comment, just an observation.

Comment (kostola) [reply executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorService.java]: I inlined it in `processEvent` to simplify the code as you suggested

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java (hunk 10 above)
Comment (kostola) [reply executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java]: There's no guarantee that the transformed event will always be a `CloudEvent` with the current implementation.

I refactored the code to include your second suggestion, which is also what I believe is the best solution at the moment:

> Prevent the user from applying a transformation to a Source processor

This allows us to finalize the first version of Event Sources quickly and think more in depth about transformations later. I honestly think source processors feature can initially be merged like this without compromising its usage.

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorsService.java (hunk 1 above)
Comment (kostola) [reply executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorsService.java]: I refactored everything to use `CloudEvent`s both in source and sink processors as you suggested.
This had two additional major benefits:
* Filtering is now consistent between both type of processors, as they always deal with a `CloudEvent` structure.
* The special `SourceAction` internal action is not needed anymore and standard `Webhook` action can be used to resolve `Source`s. Thus, I deleted it completely.

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java (hunk 7 above)
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: With the refactoring [described here](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/664#discussion_r862846024) now the event received from a source Managed Connector is wrapped in a `CloudEvent` before reaching the action invoker, so the filtering is consistent betweek source and sink processors.

I honestly have no idea about Kafka Event headers. We are totally not dealing with them now.

Imho this should be discussed separately.

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java [hunk 12]
```diff
@@ -156,11 +167,16 @@ public Processor updateProcessor(String bridgeId, String processorId, String cus
         if (!Objects.equals(existingAction, processorRequest.getAction())) {
             throw new BadRequestException("It is not possible to update the Processor's Action.");
         }
+        if (!Objects.equals(existingSource, processorRequest.getSource())) {
+            throw new BadRequestException("It is not possible to update the Processor's Source.");
+        }
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:173]: @kostola Can you please add a check in here that prevents a user from trying to update the `type` of a `Processor` between `sink` and `source`.

cheers,

Rob

File: dev/SourceProcessors.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    {
+      "key": "data.text",
+      "type": "StringContains",
+      "values": [ "hello", "hi" ]
+    }
+  ]
+}
+```
+
+## Currently implemented sources
```
Comment (robpblake) [inline dev/SourceProcessors.md:37]: @kostola We have `ACTIONS.md` in the root of the repo for the list of `Actions` we support. I think we should have something similar for the list of `Sources` that we support. Our other general user-facing docs e.g `ACTIONS.md`, `FILTERS.md` and `TRANSFORMATIONS.md` will also need updating to reflect the introduction of `Source Processors`. 

Those `.md` files are useful for the docs team who are doing their initial first iteration of customer-facing content now.

cheers,

Rob

File: manager/src/main/java/com/redhat/service/smartevents/manager/api/models/requests/ProcessorRequest.java [hunk 14]
```diff
@@ -42,13 +44,29 @@ public ProcessorRequest(String name, Action action) {
...
 
+    @JsonIgnore
+    public ProcessorType getType() {
+        if (getSource() != null) {
+            return ProcessorType.SOURCE;
+        }
+        if (getAction() != null) {
+            return ProcessorType.SINK;
+        }
+        return null;
```
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/api/models/requests/ProcessorRequest.java:67]: Wouldn't it have sense to throw  an exception in here (fail fast)?
If the processor type cannot be resolved than IMHO something is wrong and we shouldn't continue.

**Review by sutaakar — COMMENTED** [review]
One minor comment fro me above, otherwise looks good to me.
Checked just the code, will leave the complex testing on Jozef.

File: executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java (hunk 10 above)
Comment (robpblake) [reply executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java]: @kostola I agree with you. I've created https://issues.redhat.com/browse/MGDOBR-682 to capture the need for `Transformation` on an Event Source.

cheers,
Rob

**Review by lucamolteni — APPROVED** [review]

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java (hunk 7 above)
Comment (robpblake) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: @kostola @lucamolteni I'm pretty sure this will impact on the S3 Source Connector (and maybe other connectors). From what I can see of the payload shared by @lucamolteni last week, the S3 Connector is using Kafka headers to store some of the data about the event emitted from S3. So we need to understand how we're going to bind the Kafka headers used by the Connector into the wrapped `CloudEvent` so that a user can filter on them in their `Processor`. For example, the Events produced by the S3 Connector seem to have the following headers:

```
Headers: {CamelAwsS3BucketName=test-connector-1, CamelAwsS3ContentControl=null, CamelAwsS3ContentDisposition=null, CamelAwsS3ContentEncoding=null, CamelAwsS3ContentLength=59650, CamelAwsS3ContentType=image/jpeg, CamelAwsS3ETag="4721d67d808c4854581fe119a914564e", CamelAwsS3ExpirationTime=null, CamelAwsS3Key=side-monkey.jpeg, CamelAwsS3LastModified=2022-04-29T13:16:01Z, CamelAwsS3Metadata={}, CamelAwsS3ReplicationStatus=null, CamelAwsS3ServerSideEncryption=null, CamelAwsS3StorageClass=null, CamelAwsS3VersionId=null, CamelMessageTimestamp=1651238161000}
```

If I as user wanted to write a `Filter` targeting the `CamelAwsS3Key`, how would I do that? 

My instinct is that we make these headers Extension Context Attributes (https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#extension-context-attributes) and then place the event payload into the `data` element of the `CloudEvent` as we do now. So taking the S3 Connector payload as an example, we'd end up with a `CloudEvent` looking like:

```
{
  "id" : "blah",   
  "source" : "RHOSE",     
  "specversion" : "1.0",
  "camelawss3bucketname": "test-connector-1",   <-- headers
  "camelawss3contentcontrol": "null", <-- headers
  ...
  "data": {
  }
}
```
Then as a user, I can write my `Filter` as follows:

```
{
  "filters": [
    {
      "type": "StringEquals", 
      "key": "camelawss3key",
      "value": "foo"
    }
  ]
}
```

@lucamolteni Can you confirm I'm correct on the structure of the Kafka Event from S3 Connector please?

So basically rule would be:

- If Source Connector uses Kafka Headers, these are hoisted to Extension Context Attributes
- Payload from the Source Connector is preserved as-is into the `data` part of the `CloudEvent`

cheers,

Rob

cc @danielezonca

File: manager/src/main/java/com/redhat/service/smartevents/manager/api/models/requests/ProcessorRequest.java (hunk 14 above)
Comment (kostola) [reply manager/src/main/java/com/redhat/service/smartevents/manager/api/models/requests/ProcessorRequest.java]: Thanks for the hint. Done :+1:

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java (hunk 12 above)
Comment (kostola) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:176]: Good hint. I introduced an additional check on processor type just to be sure (+ tests).

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java (hunk 7 above)
Comment (kostola) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: @danielezonca @lucamolteni @robpblake I think this discussion is getting big enough to require a dedicated task...wdyt?

Comment (robpblake) [reply processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/source/SourceActionInvoker.java]: Thanks @kostola @lucamolteni 👍

**Review by sutaakar — APPROVED** [review]

File: dev/SourceProcessors.md (hunk 13 above)
Comment (kostola) [reply dev/SourceProcessors.md]: I updated all these files and added a new `SOURCES.md` file :+1:

**Review by robpblake — APPROVED** [review]
Very nice. Thank you!

File: SOURCES.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+## Parameters of a Source
+
+Each `Source` has 2 parameters to specify:
+- `type`: the type of the `Source`. This must be one of the supported `Source` types listed below
+  - Attempting to use an unknown `Source` type will result in an Error from the Bridge API.
+- `parameters`: A key/value map of configuration parameters for the `Source`
+  - Only string for the `key` and `value` of the parameters are supported.
+  - The required parameters are `Source` specific and documented in the list of supported `Sources`
+
+## Supported Source Types
```
Comment (manstis) [inline SOURCES.md:24]: @kelvah FYI.. this seems to be the only "Source Processor" we need to worry about in the UI at the moment.

File: manager/src/main/resources/db/migration/0.1/V0.1.10__update_connector.sql [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+alter table CONNECTOR
+    rename column connector_type to connector_type_id;
+
+-- add new "type" column
+alter table CONNECTOR
+    add column "type" varchar(255);
+
+-- update existing connectors "type" with 'SINK' value (the only possible one before this migration)
+update CONNECTOR
+    set "type"='SINK';
```
Comment (manstis) [inline manager/src/main/resources/db/migration/0.1/V0.1.10__update_connector.sql:11]: Would it not be safer to include `where type is null`?

**Review by manstis — APPROVED** [review]

File: manager/src/main/resources/db/migration/0.1/V0.1.10__update_connector.sql (hunk 16 above)
Comment (kostola) [reply manager/src/main/resources/db/migration/0.1/V0.1.10__update_connector.sql:11]: there can't be source connectors before merging this PR

Comment (manstis) [reply manager/src/main/resources/db/migration/0.1/V0.1.10__update_connector.sql:11]: Indeed, but there can be afterwards and this is re-deployed to an environment unless we always delete the database.

It was only a comment; I approved anyway.
